### PR TITLE
[definitely don't merge] Tendrilizing Keeps

### PR DESCRIPTION
**Pull request by ryanpbrewster**
@andrewconner @leogrim @gratimax

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala [hunk 1]
```diff
@@ -248,6 +249,10 @@ class KeepInterner @Inject() (
           } |> { keep =>
             keepRepo.save(keep)
           }
+        keepToLibraryRepo.getByKeepIdAndLibraryId(savedKeep.id.get, library.id.get) match {
+          case None => keepToLibraryRepo.save(KeepToLibrary(keepId = savedKeep.id.get, libraryId = library.id.get, keeperId = userId))
+          case Some(ktl) =>
```
Comment (maxov) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: A little bit misleading (as if we'll do something with ktl), just `case _ =>` would work

Actually, I think `if (keepToLibraryRepo.getByKeepIdAndLibraryId(savedKeep.id.get, library.id.get).isEmpty) { <logic> }` would be most clear

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala]: Haha, good suggestion. I just did exactly that. :)

File: kifi-backend/modules/common/app/com/keepit/model/KeepToLibrary.scala [hunk 2]
```diff
@@ -23,3 +26,38 @@ case class KeepToLibrary(
...
+  keepId: Id[Keep],
+  libraryId: Id[Library],
+  requesterId: Id[User]) extends KeepToLibraryRequest
+case class KeepToLibraryAttachResponse(link: KeepToLibrary)
+
+case class KeepToLibraryDetachRequest(
+  keepId: Id[Keep],
+  libraryId: Id[Library],
+  requesterId: Id[User]) extends KeepToLibraryRequest
+case class KeepToLibraryDetachResponse(dummy: Boolean = false)
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/model/KeepToLibrary.scala:63]: Should we pick between "attaching" vs "linking" a keep to a library? Or is there a difference I'm missing here? 
`link` sounds a bit generic, what about just `AddKeepToLibraryRequest` and `RemoveKeepFromLibraryRequest`?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 3]
```diff
@@ -753,7 +754,9 @@ class LibraryCommanderImpl @Inject() (
         }
 
         // Update visibility of keeps
+        // TODO(ryan): We need to find a new way of describing keep visibility. Denormalizing is no longer possible because it's different for different users
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:757]: This method should go over `KeepToLibrary` as opposed to `Keep`?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 4]
```diff
@@ -1211,6 +1215,10 @@ class LibraryCommanderImpl @Inject() (
     }
   }
 
+  // TODO(ryan): is this actually necessary anymore? Check with Léo to see if Search needs it
+  // We may not need to have this concept of "keep ownership" anymore. You can be the author of
+  // a keep, which is its own thing. You can have access to edit a keep, which is also its own
+  // thing.
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:1220]: This method was only used to transfer pre-built libraries to influencers, I'm not sure how much we need it anymore.
Regardless, I think we must retain the `userId` field on a keep, in order to identify the creator / sender. Currently only a keep's creator is allowed to edit its metadata (title, note, image), breaking this assumption would introduce more issues in our current design (who is a keep attributed to when someone other than its creator modifies it?).

File: kifi-backend/modules/common/app/com/keepit/model/KeepToLibrary.scala (hunk 2 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/KeepToLibrary.scala:63]: The verbiage in my head makes sense, as you `attach` two things together, thus generating a `link` between the two. Later you can `detach` those two things, breaking the `link`.

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/model/KeepToLibrary.scala:63]: Well you could say that of pretty much any database relationship, it
doesn't mean we always use "attach" and "link". Should we be more business
specific?

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/KeepToLibrary.scala:63]: Is your objection specifically to the word `link`?

`attach` and `detach` already take specific requests (`KeepToLibraryAttachRequest` and `KeepToLibraryDetachRequest` respectively). I like being very explicit with the request/response types, but for the method names themselves I'm not as particular.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 3 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:757]: There's a tradeoff. We could just always look up the library to find the library's `visibility`, or we could denormalize the `visibility` onto the `KeepToLibrary` pin. I imagine that the place this is going to be most important is search, so I'll let you dictate this choice.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepsCommander.scala [hunk 5]
```diff
@@ -365,32 +365,7 @@ class KeepsCommander @Inject() (
...
-    val (mainLib, secretLib) = libraryCommander.getMainAndSecretLibrariesForUser(keep.userId)
-    def getLibFromPrivacy(isPrivate: Boolean) = {
-      if (isPrivate) Some(secretLib.id.get) else Some(mainLib.id.get)
-    }
-    if (isPrivate.isDefined && isPrivate.get != keep.isPrivate) {
-      keepRepo.save(keep.copy(visibility = Keep.isPrivateToVisibility(isPrivate.get), libraryId = getLibFromPrivacy(isPrivate.get)).withTitle(updatedTitle))
-    } else {
-      keepRepo.save(keep.withTitle(updatedTitle))
-    }
-  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepsCommander.scala:392]: This code was used to migrate from the `public/private` paradigm to the Library-based one (it moves `public -> SYSTEM_MAIN` and `private -> SYSTEM_SECRET`). No longer needed.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:757]: We definitely need to denormalize visibility / orgId / userId from library
onto KeepToLibrary, just like we do today on keep :)

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala (hunk 4 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:1220]: @andrewconner

**Pull request closed without merging**